File: day_3/puzzle_1/solve.rb
# frozen_string_literal: true

require_relative '../../benchmark/benchmark'

Benchmark.run('day_3_puzzle_1') do
  def letter_to_priority(letter)
    char_code = letter.ord
    char_code >= 97 ? char_code - 96 : char_code - 38
  end

  File.open('./day_3/input.txt', 'r') do |file|
    ruck_sacks = file.read.split("\n")

    priority = ruck_sacks.inject(0) do |sum, ruck_sack|
      compartment_size  = ruck_sack.length / 2

      compartment_one   = ruck_sack.slice(0, compartment_size)
      compartment_two   = ruck_sack.slice(compartment_size, ruck_sack.length)

      compartment_one.each_char do |char|
        if compartment_two.include?(char)
          sum += letter_to_priority(char)
          break
        end
      end

      sum
    end

    priority
  end
end

File: day_3/puzzle_1/solve.py
import os
import sys
from functools import reduce

sys.path.append(
    os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
)

from benchmark.benchmark import benchmark


def main():
    def letter_to_priority(letter: str):
        char_code = ord(letter)
        return char_code - 96 if char_code >= 97 else char_code - 38

    def assess_ruck_sack(sum: int, ruck_sack: str):
        compartment_size = len(ruck_sack) // 2

        compartment_one = ruck_sack[:compartment_size]
        compartment_two = ruck_sack[compartment_size : len(ruck_sack)]

        for char in compartment_one:
            if char in compartment_two:
                sum += letter_to_priority(char)
                break

        return sum

    with open("./day_3/input.txt", "r") as file:
        ruck_sacks = file.read().splitlines()
        priority = reduce(assess_ruck_sack, ruck_sacks, 0)
        return priority


if __name__ == "__main__":
    benchmark(main, "day_3_puzzle_1")
